keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md

```diff
@@ -0,0 +1,449 @@
+(e.g., the queue is initialized with specified parameters like `LessFunc`, `SharedInformerFactory`,
+and `Option` by [factory.go#create](https://github.com/kubernetes/kubernetes/blob/f1f0183d2bbcde33024b2a05d6f39df32f11e037/pkg/scheduler/factory.go#L172)),
+this solution will not decouple them as this has no impact on the goal of custom queue,
+and such logics can be enhanced in the future on demand.
+
+Some of the items like `SchedulingQueue`, `Option` that are only used by the internal
+queue now will be exposed at [interface.go](https://github.com/kubernetes/kubernetes/blob/fb9cafc99be94a73d9b92545164dbf336bbd230a/pkg/scheduler/framework/interface.go),
+users of custom queue can access them.
+
+To implement a custom queue, what an user needs to do is to copy the codes of the current
+internal queue (e.g., scheduling_queue.go), update some logics to get the custom queue,
+and then pass the custom queue to the scheduler like the plugins. With this way, user's
+logics will override the logics of the current internal queue.
+
+The codes will be updated as following:
+
+1. Make the interface `SchedulingQueue` as a public one, so that users can implement it
+for the custom queue.
+
+2. Provide a new function `WithCustomQueue`, users can register the custom queue with this
+function like other plugins.
+
+    ```go
+         command := app.NewSchedulerCommand(
+             app.WithPlugin(coscheduling.Name, coscheduling.New),
+             ......
+       +     app.WithCustomQueue(customQueue.New),
+         )
+    ```
+
+3. The registered custom queue will be recorded by updated `Registry` struct.
+
+    ```go
+      type Registry struct {
+          Pf          map[string]PluginFactory
+          CustomQueue schedulingQueue.SchedulingQueue
+      }
+    ```
+
+4. In function `(c *Configurator) create()`, custom queue will be passed to the scheduler
+if it exists, or the current internal queue will be used.
+
+    ```go
+      - podQueue := internalqueue.NewSchedulingQueue(
+      -         lessFn,
+      +
+      + var podQueue framework.SchedulingQueue
+      + if c.registry.CustomQueue != nil {
+      +         podQueue = c.registry.CustomQueue
+      + } else {
+      +         podQueue = &internalqueue.PriorityQueue{}
+      + }
+      + podQueue.Init(...)
+    ```
+
+5. Basic inputs for queue's initialization will not be changed, a new method `Init` is added in
+the interface `SchedulingQueue`, users must implement it.
+
+    ```go
+      + func (p *PriorityQueue) Init(
+      +        lessFn framework.LessFunc,
+      +        informerFactory informers.SharedInformerFactory,
+      +        opts ...framework.Option,
+      + ) {
+    ```
+
+### Test Plan
+
+- **Unit Tests**: All core changes must be covered by unit tests.
+- **Integration Tests**: One integration test for the custom queue.
+- **Benchmark Tests**: The performance benchmark test result is same as before if custom queue is not used.
+
+### Graduation Criteria
+
```

Comment:
Describe the plan of how to roll it out here. Like phase 1, phase 2, etc.

Reply:
This document is for the solution of phase 1 (provide interface to implement custom queue), and there is no solution details for phase 2 (refine the codes, decouple the queue with some structures), should we mention the `Graduation Criteria` of phase 2 in this document?

Reply:
I think you should have *some* idea of what phase 2 would look like before we proceed.

Reply:
For phase 2, what we will do is
1. refine the codes, make the interface simple and easy to use, decouple this feature with other parts of the scheduler
2. refine some designs with user's feedback

Update the message as following:
```
 #### Beta -> GA Graduation

+- Only necessary interface and structs are exposed to users.
 - Allowing time for feedback to ensure that the new interface sufficiently expresses users requirements.
```

Reply:
For phase 2, what we will do is

refine the codes, make the interface simple and easy to use, decouple this feature with other parts of the scheduler
refine some designs with user's feedback
Update the message as following: